src/Compilers/Core/RebuildTest/DeterministicKeyBuilderTests.cs

```diff
@@ -0,0 +1,727 @@
+        {
+            var syntaxTree = ParseSyntaxTree("", fileName: "test", SourceHashAlgorithm.Sha256, (TParseOptions)parseOptions);
+            var compilation = CreateCompilation(syntaxTrees: new SyntaxTree[] { syntaxTree });
+            var property = GetJsonProperty(compilation.GetDeterministicKey(), "compilation.syntaxTrees");
+            var trees = (JArray)property.Value;
+            var obj = (JObject)trees[0];
+            return (JObject)(obj.Property("parseOptions")?.Value!);
+        }
+
+        protected JObject GetReferenceValue(MetadataReference reference)
+        {
+            var compilation = CreateCompilation(syntaxTrees: new SyntaxTree[] { }, references: new[] { reference });
+            var property = GetJsonProperty(compilation.GetDeterministicKey(), "compilation.references");
+            var expectedMvid = DeterministicKeyBuilder.GetGuidValue(reference.GetModuleVersionId());
+
+            var array = (JArray)property.Value;
+            foreach (var item in array!.Values<JObject>())
+            {
+                if (item?.Value<string>("mvid") == expectedMvid)
+                {
+                    return item;
+                }
+            }
+
+            Assert.True(false, $"Could not find reference with MVID {expectedMvid}");
+            throw null!;
+        }
+
+        protected static JArray GetSyntaxTreeValues(Compilation compilation, ImmutableArray<KeyValuePair<string, string>> pathMap = default)
+        {
+            var property = GetJsonProperty(compilation.GetDeterministicKey(pathMap: pathMap), "compilation.syntaxTrees");
+            return (JArray)property.Value; ;
+        }
+
+        protected static JArray GetAdditionalTextValues(Compilation compilation, ImmutableArray<AdditionalText> additionalTexts, ImmutableArray<KeyValuePair<string, string>> pathMap = default)
+        {
+            var property = GetJsonProperty(compilation.GetDeterministicKey(additionalTexts: additionalTexts, pathMap: pathMap), "additionalTexts");
+            return (JArray)property.Value;
+        }
+
+        private protected JObject GetEmitOptionsValue(
+            EmitOptions emitOptions,
+            ImmutableArray<KeyValuePair<string, string>> pathMap = default,
+            DeterministicKeyOptions options = default)
+        {
+            var compilation = CreateCompilation(new SyntaxTree[] { });
+            var key = compilation.GetDeterministicKey(
+                emitOptions: emitOptions,
+                pathMap: pathMap,
+                options: options);
+            var property = GetJsonProperty(key, "emitOptions");
+            return (JObject)property.Value;
+        }
+
+        protected static string GetChecksum(SourceText text)
+        {
+            var checksum = text.GetChecksum();
+            var builder = PooledStringBuilder.GetInstance();
+            DeterministicKeyBuilder.EncodeByteArrayValue(checksum.AsSpan(), builder);
```

Comment:
Could use the `EncodeByteArrayValue` method that takes a single ReadOnlySpan parameter (and no builder)